server/hotrod/src/main/scala/org/infinispan/server/hotrod/Decoder2x.scala

```diff
@@ -397,7 +419,28 @@ object Decoder2x extends AbstractVersionedDecoder with ServerConstants with Log
       }
    }
 
-   override def customReadValue(header: HotRodHeader, buffer: ByteBuf, cache: Cache): AnyRef = null
+   override def customReadValue(decoder: HotRodDecoder, h: HotRodHeader,
+       hrCtx: CacheDecodeContext, buffer: ByteBuf, cache: Cache): AnyRef = {
+      h.op match {
+         case PutAllRequest =>
+            var map = hrCtx.putAllMap
+            if (map == null) {
+              map = new HashMap[Bytes, Bytes]
+              hrCtx.putAllMap = map
+            }
+            for (i <- map.size until hrCtx.params.valueLength) {
+              val k = readRangedBytes(buffer)
+              val v = readRangedBytes(buffer)
+              map.put(k, v)
+              // We check point after each read entry
+              decoder.checkpoint
```

Comment:
Revisited the code and indeed it looks good. The caching is indeed being done to avoid re-reading the read entries. Sorry I missed these details first time around.